content/blog/eng/from-0-to-infinite-the-final-keyword.md

```diff
@@ -0,0 +1,338 @@
+But the problem is that these principles are difficult to apply, they almost always lead to writing more code. 
+Not always this is possible, not always this is accepted. 
+
+Regarding final, in my opinion, one of the principle that best represents it is LSP.
+
+Let's see what **Liskov principle** says in more detail. 
+The canonical statement says the following:
+
+> "you should always be able to substitute the parent class with its derived class without any undesirable behaviour"
+
+> The concept of this principle was introduced by Barbara **Liskov** in a 1987 conference keynote and later published in a paper together with Jannette Wing in 1994. 
+
+The more precise definition:
+>“if S is a subtype of T, then objects of type T in a program may be replaced with objects of type S without altering any of >the desirable properties of that program.”
+
+Link to orginal paper: https://dl.acm.org/doi/pdf/10.1145/62138.62141
+
+> Simplifying the following could be good guidelines:
+
+> - Don’t stricter more validation rules on input parameters than those implemented by the parent class.
+> - Objects should be replaceable with instances of their subtypes without altering the correctness of that program
+> - Not add restriction in an overridden method if that restriction doesn't exist in baseline implementation
+
+I have used it like as example because is specifically for the design of hierarchies.
+
+I do not want to write an article on how to respect the development according to **Liskov** in the net there is a lot of literature. I would just like to emphasize that building a hierarchy of objects is not easy, that more the hierarchy is deep then more we are faced with design problems and **final** helps use to prevent this. 
+
+For this reason, before creating subclasses we should ask ourselves questions such as these or similar:
+
+- is my derived class of the same type as the base class?
+- I can use the derived class of base where requireds a base class without having "strange" behaviors in the code at runtime?
+- is my base class prepared to be derived?
+- the enhancements made by the subclass are primarily additive?
+
+If the answers to these questions are negative, using inheritance in this use case it could lead us to create complex and unmanageable hierarchies. 
+
+Now some basic example of wrong use of inheritance:
+```php
+use App\Domain\Model;
+class Stack extends ArrayList {
+    public push(Object value) { … }
+    public Object pop() { … }
+}
+```
+
+Why ? The Stack class inheriting from ArrayList will have a lot of additional methods not related to the concept of ArrayList (e.g. push and pop).
+
+In the previous example two main errors have been made, first error a Stack IS NOT an Array, second problem, is the cross-domain inheritance relationship. Simply, our domain classes must use implementations not to inherit them, in our example, stack was a concept of domain (Focusing only on reusing code can be a problem).
```

Comment:
In the previous example two main errors have been made: 
1. Stack IS NOT an Array. 
2. The cross-domain inheritance relationship.
Our domain classes must use implementations not to inherit them. In the example, stack was a concept of domain (focusing on reusing code only can be a problem).